src/System.Text.Json/src/System/Text/Json/JsonElement.cs

```diff
@@ -0,0 +1,372 @@
+﻿using System.Collections.Generic;
+using System.Diagnostics;
+
+namespace System.Text.Json
+{
+    public struct JsonElement
+    {
+        private readonly JsonDocument _parent;
+        private readonly int _idx;
+
+        internal JsonElement(JsonDocument parent, int idx)
+        {
+            _parent = parent;
+            _idx = idx;
+        }
+
+        public JsonTokenType Type => _parent?.GetJsonTokenType(_idx) ?? JsonTokenType.None;
+
+        public JsonElement this[int index]
+        {
+            get
+            {
+                if (_parent == null)
+                {
+                    throw new InvalidOperationException();
```

Comment:
Do we gain anything perf-wise if we were to move this out with the ThrowHelper pattern?

Also, I assume error messages will be added later?

Reply:
I moved all of these checks "am I default(JsonElement)?" to a helper CheckValidInstance, and all of the disposed checks in JsonDocument to CheckNotDisposed and all of the "am I the right type?" checks in JsonDocument to CheckExpectedType... and the performance is either unchanged or worse.  (within overlapping sigmas, but the "after" had a higher mean when the JsonObject baseline got a lower mean).

But I guess those changes are fine, especially if you think we want to be more helpful on "don't call members on the default JsonElement" case by using a non-default string.

Or maybe I don't understand what the actual ThrowHelper pattern is.

Reply:
Did you try a throw helper after adding a message string? @benaadams previously saw a benefit (at least in asm) here: https://github.com/dotnet/corefxlab/pull/2366#discussion_r198356048

Also, from @AndyAyersMS,
https://github.com/dotnet/corefx/pull/33216#discussion_r230576566

> Methods that conditionally throw exceptions on some paths can still be inlined, as can methods with complex flow where every path ultimately throws.

> Simple methods that unconditionally throw won't be inlined w/o AggressiveInlining.

> The benefits of the throw helper pattern are that it can save code size if the same exception can be thrown from multiple places within a method or across multiple methods, and that it also may reduce some of the time and space overhead the throwing path may impose on non-throwing paths in a method (say the method prolog).

> Using a throw helper for just one site in one method may not provide any benefit -- you'd have to look at the non-exceptional codegen paths with and without a helper to see.